tokio/src/sync/mpsc/chan.rs

```diff
@@ -239,6 +239,10 @@ impl<T, S: Semaphore> Rx<T, S> {
         self.inner.notify_rx_closed.notify_waiters();
     }
 
+    pub(crate) fn is_closed(&self) -> bool {
+        self.inner.semaphore.is_closed()
```

Comment:
IMO, `is_closed()` should return `true` if the next call to `recv()` returns `None`. For this to be true, the closed flag needs to be set **and** the channel needs to be empty, which is a bit harder to check.

Maybe we should add an `is_empty()` method as well?

Second, I don't actually remember and would have to look at the code, but isn't the sempahore.is_closed() signaling that the **receiver** half dropped? To verify, one would need to read the `recv()` path to see how it checks for a closed channel.

Reply:
Agree here as well. I think it would be easier here if we added test cases which demonstrated this behavior.

Reply:
> Second, I don't actually remember and would have to look at the code, but isn't the sempahore.is_closed() signaling that the **receiver** half dropped? To verify, one would need to read the `recv()` path to see how it checks for a closed channel.

I believe that's correct, FWIW.

Reply:
Soo...I was taking a look at the code and noticed this bit, esp the Empty condition
```
  pub(crate) fn try_pop(&mut self, tx: &Tx<T>) -> TryPopResult<T> {
      let tail_position = tx.tail_position.load(Acquire);
      let result = self.pop(tx);

      match result {
          Some(block::Read::Value(t)) => TryPopResult::Ok(t),
          Some(block::Read::Closed) => TryPopResult::Closed,
          None if tail_position == self.index => TryPopResult::Empty,
          None => TryPopResult::Busy,
      }
  }
```
If I'm reading it correctly, it basically means that if pop doesn't return a value and if tail and head are the same, it means we have exhausted all the blocks we can read. If we can guarantee that semaphore is closed, then I believe simply checking these indices should do the trick. I have a very rough POC which seems to be working for the simplest case below:
```
#[tokio::main]
async fn main() {
    let (tx, mut rx) = mpsc::channel(42);
    
    assert!(!rx.is_closed());

    tx.send(100).await.unwrap();
    tx.send(200).await.unwrap();

    rx.close();
    assert!(!rx.is_closed());
    assert_eq!(rx.recv().await, Some(100));

    assert!(!rx.is_closed());
    assert_eq!(rx.recv().await, Some(200));

    assert!(rx.is_closed())
}
```
Please LMK if I'm completely shooting in the dark here or this looks like a start? If you believe this is too complicated for a first time contribution I can back off but happy to work on this if you folks have enough patience. :)

Reply:
The implementation with indices is probably reasonable.